server/app/controllers/api/ProgramApplicationsApiController.java

```diff
@@ -229,10 +285,76 @@ private Instant parseParamDateToInstant(String paramName, String paramDate) {
     }
   }
 
-  private RowIdSequentialAccessPaginationSpec createPaginationSpec(
-      ApiPaginationTokenPayload apiPaginationTokenPayload) {
-    return new RowIdSequentialAccessPaginationSpec(
-        apiPaginationTokenPayload.getPageSpec().getPageSize(),
-        Long.valueOf(apiPaginationTokenPayload.getPageSpec().getOffsetIdentifier()));
+  /**
+   * Resolves the lifecycle stage from either the revision state parameter or pagination token.
+   * Ensures consistency between pagination token and query parameter when both are present.
+   */
+  private ImmutableList<LifecycleStage> resolveRevisionStateParam(
+      Optional<ApiPaginationTokenPayload> apiPaginationTokenPayload,
+      Optional<String> revisionStateParam) {
+    // Defaults to all submitted application if no revision state is provided
+    if (revisionStateParam.isEmpty()) {
+      return ImmutableList.of(LifecycleStage.ACTIVE, LifecycleStage.OBSOLETE);
+    }
+
+    ImmutableList<LifecycleStage> tokenLifecycleStages =
+        apiPaginationTokenPayload
+            .map(
+                (token) ->
+                    extractLifecycleStagesFromPaginationToken(token, revisionStateParam.get()))
+            .orElse(ImmutableList.of());
+    ImmutableList<LifecycleStage> queryLifecycleStages = getLifecycleStages(revisionStateParam);
+
+    if (tokenLifecycleStages.equals(queryLifecycleStages)) {
+      return tokenLifecycleStages;
+    }
+
+    if (!tokenLifecycleStages.isEmpty() && !queryLifecycleStages.isEmpty()) {
+      throw new BadApiRequestException(
+          "Request parameters must match pagination token: " + revisionStateParam);
+    }
+
+    return !tokenLifecycleStages.isEmpty() ? tokenLifecycleStages : queryLifecycleStages;
+  }
+
+  /**
+   * Extracts the lifecycle stage from a pagination token's request specification.
+   *
+   * @param apiPaginationTokenPayload The pagination token containing previous request context
+   * @param paramName The parameter name to look for in the token's request spec
+   * @return Optional LifecycleStage if found in token, empty otherwise
+   */
+  private ImmutableList<LifecycleStage> extractLifecycleStagesFromPaginationToken(
+      ApiPaginationTokenPayload apiPaginationTokenPayload, String paramName) {
+    Map<String, String> requestSpec = apiPaginationTokenPayload.getRequestSpec();
+
+    if (!requestSpec.containsKey(paramName)) {
+      return ImmutableList.of();
+    }
+
+    String stagesString = requestSpec.get(REVISION_STATE_PARAM_NAME);
+    return Arrays.stream(stagesString.split(","))
+        .map(String::trim)
+        .map(LifecycleStage::valueOf)
+        .collect(ImmutableList.toImmutableList());
+  }
```

Comment:
(blocking) This is mixing the public `RevisionState` with the internal `LifecycleStage`. The `RevisionState` should be encoded into request spec.